lib/commit-lint.ts

```diff
@@ -0,0 +1,134 @@
+    message: string;
+}
+
+/*
+ * Simple class to drive lint tests on commit messages.
+ */
+export class LintCommit {
+    private failed: boolean;
+    private lines: string[];
+    private patch: IPRCommit;
+    private messages: string[] = [];
+
+    public constructor(patch: IPRCommit) {
+        this.failed = false;
+        this.lines =  patch.message.split("\n");
+        this.patch = patch;
+    }
+
+    /*
+     * Linter method to run checks on the commit message.
+     * @param {IPRCommit} the patch to be checked
+     *
+     * The lintings are in methods called from here.  If
+     * a lint check is too severe to continue, it will throw
+     * an error.
+     */
+    public async lint(): Promise<ILintError | void> {
+
+        // Basic test before all others
+
+        if (await this.commitViable()) {
+            await Promise.all([
+                this.commitMessageLength(),
+                this.lowerCaseAfterPrefix(),
+                this.signedOffBy(),
+                this.moreThanAHyperlink(),
+            ]);
+        }
+
+        if (this.messages.length) {
+            return { checkFailed: this.failed,
+                     message: `There ${this.messages.length > 1 ?
+                     "are issues" : "is an issue"} in commit ${
+                     this.patch.commit}:\n${this.messages.join("\n")}`,
+                };
+        }
+    }
+
+    private addMessage(message: string, failed: boolean = false): void {
+        this.messages.push(message);
+        if (failed) {
+            this.failed = true;
+        }
+    }
+
+    // Test for a minimum viable commit message.
+    // - the body of the commit message should not be empty
+
+    private async commitViable(): Promise<boolean> {
+        if (this.lines.length < 3) {
+            this.addMessage("Commit checks stopped - the message is too short",
+                            true);
+            return false;
+        }
+
+        return true;
+    }
+
+    // More tests of the commit message structure.
+    // - the first line should not exceed 76 characters
+    // - the first line should be followed by an empty line
+
+    private async commitMessageLength(): Promise<void> {
+        if (this.lines[0].length > 76) {
+            this.addMessage("First line of commit message is too long " +
+                            `(> 76 columns): ${this.lines[0]}`,
+                            true);
+        }
+
+        if (this.lines[1].length) {
+            // tslint:disable-next-line:max-line-length
+            this.addMessage("The first line must be separated from the rest by an empty line", true);
+        }
+    }
+
+    // Verify if the first line starts with a prefix (e.g. tests:), it continues
+    // in lower-case
+
+    private async lowerCaseAfterPrefix(): Promise<void> {
+        const match = this.lines[0].match(/^([a-z]+)+?:.*?([A-Z])/);
+
+        if (match && typeof(match[2]) === "string" ) {
```

Comment:
I don't think we have to verify that `match[2]` is of type `string` here.

Reply:
I think that was left over - removed with no effect on the unit tests.

Reply:
Excellent!